File: ios-native/FretboardHeatmap/HeatmapLogic.swift
//
//  HeatmapLogic.swift
//  FretboardHeatmap
//
//  Created by Sean Avery Suguitan on 2026-01-20.
//

import SwiftUI

/// Calculate and display fret positions on the fretboard
struct HeatmapLogic: View {
    let selectedRoot: String?
    let topMenu: menuChoice?
    let selectedDropdownOption: String
    let noteLabels: Bool
    let frets: [CGFloat]
    
    // MARK: Display frets
    var body: some View {
        VStack(spacing: 0) {
            ForEach(0..<6, id: \.self) { string in
                Spacer() // Center Vertically
                
                // Display frets
                Color.clear
                    .frame(height: 1.0)
                    .overlay(alignment: .leading) {
                        heatmapFrets(for: string)
                    }
            }
            Spacer() // Center Horizontally
        }
    }
    
    // MARK: Prepare frets for the heatmap
    /// Prepare frets for the heatmap
    /// - Parameters:
    ///    - string: The current string
    @ViewBuilder
    private func heatmapFrets(for string: Int) -> some View {
        if let root = selectedRoot, let topMenu = topMenu {
            // selected root frets
            let positions = FretPositions.getFretMap(topMenu: topMenu, dropdownChoice: selectedDropdownOption, root: root)
            
            // center frets
            ForEach(positions[string] ?? [], id: \.self) { fret in
                
                if fret > 0 && fret <= frets.count {
                    centerNotes(root: root, string: string, fret: fret)
                }
            }
        }
    }

    // MARK: Calculate center of fret and label positions
    /// Calculate center of fret and label positions
    /// - Parameters:
    ///    - root: The current root
    ///    - string: The current string
    ///    - fret: The current fret
    @ViewBuilder
    private func centerNotes(root: String, string: Int, fret: Int) -> some View {
        // center of fret
        // total width from fret 1 to current fret
        let woodDistance: CGFloat = frets.prefix(fret).reduce(0.0, +)
        
        // account for fret 1 to current fret wire widths
        let wireOffset: CGFloat = CGFloat(fret) * 3.0
        
        // - half of current frets rightmost edge
        let thisFretWidth: CGFloat = frets[fret - 1]
        let centerOfWood: CGFloat = (woodDistance + wireOffset) - (thisFretWidth / 2.0)
        
        // center of labels
        NoteCircle(
            root: root,
            string: string,
            fret: fret,
            topMenu: topMenu,
            dropdown: selectedDropdownOption,
            showLabels: noteLabels
        )
        
        .offset(x: centerOfWood - 3.5)
        .transition(.opacity.combined(with: .scale))
    }
}

// MARK: Style frets
/// Determine the notes color and label
struct NoteCircle: View {
    let nonRootNoteLabel: String
    let rootNoteLabel: Bool
    let noteLabels: Bool
    
    // labels
    init(root: String, string: Int, fret: Int, topMenu: menuChoice?, dropdown: String, showLabels: Bool) {
        self.rootNoteLabel = RootNotePositions.check(root: root, string: string, fret: fret)
        self.nonRootNoteLabel = FretLabels.getLabels(topMenu: topMenu, root: root, dropdownChoice: dropdown, string: string, fret: fret)
        self.noteLabels = showLabels
    }

    var body: some View {
        Circle()
            // size
            .fill(rootNoteLabel ? Color.red : Color.blue)
            .frame(width: 24.0, height: 24.0)
        
            // label
            .overlay {
                if noteLabels {
                    Text(nonRootNoteLabel)
                        .font(.system(size: 15.0, weight: .bold))
                        .foregroundColor(.white)
                }
            }
    }
}
#Preview {
    ContentView()
}

File: android-native/FretboardHeatmap/app/src/main/java/com/example/fretboardheatmap/HeatmapLogic.kt
package com.example.fretboardheatmap

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInHorizontally
import androidx.compose.animation.slideOutHorizontally
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * Display the heatmap dots
 *
 * Uses [AnimatedContent] with [Triple] state to animate heatmap dots whenever targetState parameters change
 *
 * @param selectedRoot current selected root from [BottomMenuArea]
 * @param topMenu current [TopMenuChoice] ("CHORDS" or "SCALES")
 * @param selectedDropdownOption current dropdown choice
 * @param noteLabels current visibility state for heatmap dot labels
 */
@Composable
fun HeatmapLogic(
    selectedRoot: String?,
    topMenu: TopMenuChoice?,
    selectedDropdownOption: String,
    noteLabels: Boolean
) {
    if (selectedRoot == null || topMenu == null) return

    // animations for heatmap dots
    // parameters handle state when transitioning between keys
    AnimatedContent(
        targetState = Triple(selectedRoot, selectedDropdownOption, topMenu),
        transitionSpec = {
            val duration = 280

            slideInHorizontally( // new key
                animationSpec = tween(duration),
                initialOffsetX = { -it }
            ) + fadeIn(animationSpec = tween(duration)) togetherWith
                    slideOutHorizontally( // old key
                        animationSpec = tween(duration),
                        targetOffsetX = { it }
                    ) + fadeOut(animationSpec = tween(duration))
        },
        label = "FretboardTransition"
    ) { (currentRoot, currentDropdown, currentMenu) ->
        Box(modifier = Modifier
            .fillMaxWidth()
            .height(265.dp)) {
            Column(
                modifier = Modifier.fillMaxSize(),
                verticalArrangement = Arrangement.SpaceEvenly
            ) {
                Spacer(modifier = Modifier.weight(1f))

                // display dots on each string
                repeat(6) { stringIndex ->
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(24.dp),
                        contentAlignment = Alignment.CenterStart
                    ) {
                        HeatmapFrets(
                            string = stringIndex,
                            selectedRoot = currentRoot,
                            topMenu = currentMenu,
                            selectedDropdownOption = currentDropdown,
                            noteLabels = noteLabels
                        )
                    }
                    Spacer(modifier = Modifier.weight(1f)) // center horizontally
                }
            }
        }
    }
}

/**
 * Map heatmap dot positions
 *
 * Uses [FretPositions] to get dot positions
 * Uses [CenterNotes] + [GuitarSpecs.frets] to center dots inside frets
 *
 * @param string the current string
 * rest of params defined in parent [HeatmapLogic]
 */
@Composable
private fun HeatmapFrets(
    string: Int,
    selectedRoot: String,
    topMenu: TopMenuChoice,
    selectedDropdownOption: String,
    noteLabels: Boolean
) {
    // get fret positions and finger numbers
    val stringPositions = remember(topMenu, selectedDropdownOption, selectedRoot, string) {
        FretPositions.getFretMap(topMenu, selectedDropdownOption, selectedRoot)[string] ?: emptyList()
    }


    // center each fret on current string
    Row(modifier = Modifier.fillMaxSize()) {
        Spacer(modifier = Modifier.width(10.dp))

        // account for current fret thickness
        GuitarSpecs.frets.forEachIndexed { index, fretWeight ->
            val currentFretNumber = index + 1

            Box(
                modifier = Modifier
                    .weight(fretWeight.value)
                    .fillMaxHeight(),
                contentAlignment = Alignment.Center
            ) {
                if (stringPositions.contains(currentFretNumber)) {
                    CenterNotes(
                        root = selectedRoot,
                        string = string,
                        fret = currentFretNumber,
                        topMenu = topMenu,
                        dropdown = selectedDropdownOption,
                        showLabels = noteLabels
                    )
                }
            }
            Spacer(modifier = Modifier.width(3.dp))
        }
    }
}

/**
 * Calculate center of heatmap dots and label positions
 *
 * Aligns [NoteCircle] inside center of current fret
 *
 * @param fret the current fret
 * @param dropdown the selected dropdown choice
 * @param showLabels current state of labels button
 * rest of params defined in [HeatmapFrets] and [HeatmapLogic]
 */
@Composable
private fun CenterNotes(
    root: String,
    string: Int,
    fret: Int,
    topMenu: TopMenuChoice,
    dropdown: String,
    showLabels: Boolean
) {
        // because parent box(HeatmapFrets) has contentAlignment.Center, alignment math used in Swift not needed
        // NoteCircle gets auto centered in current fret
        NoteCircle(
            root = root,
            string = string,
            fret = fret,
            topMenu = topMenu,
            dropdown = dropdown,
            showLabels = showLabels
        )
}

/**
 * Determine the heatmap dots color and labels display
 *
 * Highlights the current [root]
 * Displays correct [FretLabels]
 *
 * params defined in [CenterNotes], [HeatmapFrets] and [HeatmapLogic]
 */
@Composable
fun NoteCircle(
    root: String,
    string: Int,
    fret: Int,
    topMenu: TopMenuChoice,
    dropdown: String,
    showLabels: Boolean
) {
    val isRoot = remember(root, string, fret) {
        NoteAlphabet.getNoteName(string, fret) == root
    }

    val labelText = remember(topMenu, root, dropdown, string, fret) {
        FretLabels.getLabels(topMenu, root, dropdown, string, fret)
    }

    Box( // dot size
        modifier = Modifier
            .size(24.dp)
            .background(if (isRoot) Color.Red else Color(0xFF208FFA), shape = CircleShape),
        contentAlignment = Alignment.Center
    ) { // labels
        if (showLabels && labelText.isNotEmpty()) {
            Text(
                text = labelText,
                color = Color.White,
                fontSize = 12.sp,
                fontWeight = FontWeight.Bold
            )
        }
    }
}